File: codes/剑指 Offer II 016. 不含重复字符的最长子字符串.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> cmap;
        int ans = 0;
        int len = 0;
        for(int i=0; i<s.size(); i++) {
            char c = s[i];
            if(cmap.find(c) == cmap.end()) {
                len++;
                cmap[c] = i;
            } else {
                auto lastpos = cmap[c];
                if(i-lastpos > len) {
                    len++;
                } else {
                    len = i-lastpos;
                }
                cmap[c] = i;
            }
            ans = max(ans, len);
        }
        return ans;
    }
};

File: codes/剑指 Offer II 016. 不含重复字符的最长子字符串.go
func lengthOfLongestSubstring(s string) int {
	cmap := make(map[rune]int)
	len := 0
	ans := 0
	for i, c := range s {
		if _, ok := cmap[c]; !ok {
			len++
		} else {
			if i-cmap[c] > len {
				len++
			} else {
				len = i - cmap[c]
			}
		}
		cmap[c] = i
		ans = max(ans, len)
	}
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}